lucene/core/src/java/org/apache/lucene/search/QueryVisitor.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import org.apache.lucene.index.Term;
+
+/**
+ * Interface to allow recursion through a query tree
+ *
+ * @see Query#visit(QueryVisitor)
+ */
+public interface QueryVisitor {
```

Comment:
I think this should have `@FunctionalInterface` annotation or alternatively `extends Consumer<Term>`?

Reply:
I've made it a functional interface, but I'm now wondering if this would be better done as an abstract class with all default implementations?  For example, if we want to add leaf methods that accept Automata, or doc values ranges, or points, or similar - it's targeted at terms currently, but that seems overly restrictive.  Plus an abstract class allows you to call super methods, whereas now you always have to remember to deal with eg `MUST_NOT` clauses if you override `getSubVisitor()`